app/controllers/declaration/WarehouseDetailsController.scala

```diff
@@ -0,0 +1,90 @@
+import javax.inject.Inject
+import models.requests.JourneyRequest
+import models.{DeclarationType, ExportsDeclaration, Mode}
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.cache.ExportsCacheService
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import views.html.declaration.warehouse_details
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class WarehouseDetailsController @Inject()(
+  authenticate: AuthAction,
+  journeyType: JourneyAction,
+  navigator: Navigator,
+  override val exportsCacheService: ExportsCacheService,
+  mcc: MessagesControllerComponents,
+  warehouseDetailsPage: warehouse_details
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with I18nSupport with ModelCacheable {
+
+  import forms.declaration.WarehouseDetails._
+
+  def displayPage(mode: Mode): Action[AnyContent] = (authenticate andThen journeyType) { implicit request =>
+    request.cacheModel.locations.warehouseIdentification match {
+      case Some(data) => Ok(warehouseDetailsPage(mode, form().fill(data)))
+      case _          => Ok(warehouseDetailsPage(mode, form()))
+    }
+  }
+
+  def saveWarehouse(mode: Mode): Action[AnyContent] = (authenticate andThen journeyType).async { implicit request =>
+    form()
+      .bindFromRequest()
+      .fold(
+        (formWithErrors: Form[WarehouseDetails]) => Future.successful(BadRequest(warehouseDetailsPage(mode, formWithErrors))),
+        form => {
+          val nextStep = request.cacheModel.`type` match {
+            case DeclarationType.STANDARD | DeclarationType.SUPPLEMENTARY =>
+              controllers.declaration.routes.DepartureTransportController.displayPage(mode)
+            case DeclarationType.SIMPLIFIED =>
+              controllers.declaration.routes.BorderTransportController.displayPage(mode)
+          }
+          updateCache(form)
+            .map(_ => navigator.continueTo(nextStep))
+        }
+      )
+  }
+
+  private def updateCache(formData: WarehouseDetails)(implicit request: JourneyRequest[AnyContent]): Future[Option[ExportsDeclaration]] =
+    updateExportsDeclarationSyncDirect { model =>
+      val warehouseDetails = model.locations.warehouseIdentification
+        .map(
+          dbWarehouseDetails =>
+            WarehouseDetails(
+              identificationNumber = dbWarehouseDetails.identificationNumber,
+              supervisingCustomsOffice = formData.supervisingCustomsOffice,
+              inlandModeOfTransportCode = formData.inlandModeOfTransportCode
+          )
+        )
+        .getOrElse(
+          WarehouseDetails(
+            supervisingCustomsOffice = formData.supervisingCustomsOffice,
+            inlandModeOfTransportCode = formData.inlandModeOfTransportCode
+          )
+        )
+      model.copy(locations = model.locations.copy(warehouseIdentification = Some(warehouseDetails)))
```

Comment:
please move update/lens to model as method

Reply:
Good point. We are reusing the model and the next story we will have to assess our approach as we might find the limitations of our use. If so we will then address this concern there. THank you